File: Week_3/mergesort.pl
#!/usr/bin/perl

use strict;
use warnings;

use List::Util qw( shuffle );

my @test = shuffle(1 .. 10);
mergesort(\@test, (scalar @test - 1), 0);
print "@test\n";

sub mergesort {
  my ($nums, $high, $low) = @_;

  if ($high <= $low) {
    return;
  }

  my $mid = int(($high + $low) / 2);

  mergesort($nums, $mid, $low);
  mergesort($nums, $high, $mid + 1);
  merge($nums, $high, $mid, $low);
}

sub merge {
  my ($nums, $high, $mid, $low) = @_;

  my @copy;
    for (0 .. (scalar @$nums - 1)) {  # make a copy so we can sort in place
      $copy[$_] = $nums->[$_];
    }

  my $i = $low;  # pointer to start of low subarray
  my $j = $mid + 1;  # pointer to start of high subarray

  for (my $k = $low; $k <= $high; $k++) {   # go through the whole list
    if ($i > $mid) {   # low is exhausted. only do high
      $nums->[$k] = $copy[$j++];
    } elsif ($j > $high) {  # high is empty. only do low
      $nums->[$k] = $copy[$i++];
    } elsif ($copy[$j] < $copy[$i]) {  # high subarray has the lower first element
      $nums->[$k] = $copy[$j++];
    } else {  # low subarray has the lower first element (or they are the same - stable sort)
      $nums->[$k] = $copy[$i++];
    }
  }
}

sub pretty_print {
  my $nums = shift;

  for (@$nums) {
    print "$_ ";
  }
  print "\n";
}

File: Week_3/merge_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SIZE 25

void merge_sort(int nums[], int length, int low, int high);
static void swap(int *x, int *y);
static void insertion_sort(int nums[], int low, int high);
static void merge (int nums[], int length, int low, int mid, int high);
static void print_nums (int nums[], int length);

int main(void)
{

  time_t t;
  srand((unsigned) time(&t));

  int nums[SIZE], i;
  for (i = 0; i < SIZE; i++) {
    nums[i] = rand() % 100;
  }

  int length = sizeof(nums) / sizeof(nums[0]);

  print_nums(nums, length);
  merge_sort(nums, length, 0, (length - 1));
  print_nums(nums, length);

  return 0;
}

void merge_sort (int nums[], int length, int low, int high)
{
  if (high <= low + 5) {
    insertion_sort(nums, low, high);
    return;
  }

  int mid = low + ((high - low) / 2);

  merge_sort(nums, length, low, mid);
  merge_sort(nums, length, (mid + 1), high);

  merge(nums, length, low, mid, high);

}

static void merge (int nums[], int length, int low, int mid, int high)
{

  int copy[length], i;

  for (i = 0; i < length; i++) {
    copy[i] = nums[i];
  }

  int first = low;
  int second = mid + 1;
  int k;

  for (k = low; k <= high; k++) {
    if (first > mid) {
      nums[k] = copy[second++];
    } else if (second > high) {
      nums[k] = copy[first++];
    } else if (copy[second] < copy[first]) {
      nums[k] = copy[second++];
    } else {
      nums[k] = copy[first++];
    }
  }
}

static void insertion_sort (int nums[], int low, int high)
{
  int i, j;

  for (i = low; i <= high; i++) {
    for (j = i; j > 0; j--) {
      if (nums[j] < nums[j - 1]) {
        swap(&(nums[j]), &(nums[j - 1]));
      }
    }
  }
}

static void swap (int *x, int *y)
{
  int temp = *x;
  *x = *y;
  *y = temp;
}

static void print_nums (int nums[], int length)
{
  int i;

  for (i = 0; i < length; i++) {
    printf("%d ", nums[i]);
  }
  printf("\n");
}
